### ISSUE TITLE
Petastrom fails with big datasets

### ISSUE THREAD

**Issue by eliorc** [uber/petastorm#241]
Using the code from the repo Github main page as reference, my code looks like follows:

```
spark = SparkSession.builder.config('spark.driver.memory', '10g').master('local[4]').getOrCreate()
sc = spark.sparkContext

with materialize_dataset(spark=spark, dataset_url='file:///opt/data/hello_world_dataset',
                         schema=MySchema, row_group_size_mb=256):

    logging.info('Building RDD...')
    rows_rdd = sc.parallelize(ids)\
        .map(row_generator)\  # Generator that yields lists of examples
        .flatMap(lambda x: dict_to_spark_row(MySchema, x))

    logging.info('Creating DataFrame...')
    spark.createDataFrame(rows_rdd, MySchema.as_spark_schema()) \
        .coalesce(10) \
        .write \
        .mode('overwrite') \
        .parquet('file:///opt/data/hello_world_dataset') 
```

Now the RDD code executes successfully but fails only the `.createDataFrame` call with the following error:

> _pickle.PicklingError: Could not serialize broadcast: OverflowError: cannot serialize a string larger than 4GiB

This is my first experience with Spark, so I can't really tell if this error originates in Spark or Petastorm.

Looking through other solutions to this error (in respect to Spark, not Petastorm) I saw that it might have to do with the pickling protocol, but I can't confirm that, neither did I find a way of altering the pickling protocol.

How could I avoid this error?

**Comment by rgruener** [comment]
This is definitely a spark problem. I believe that what is happening is that when spark is serializing the data between the python and java processes it is failing since you have a field/row that is larger than 4Gb. What version of spark are you using? How big are your fields/rows?

**Comment by selitvin** [comment]
One more thing to try is to explicitly set number of partitions to a larger number. It all depends on the number of records and the size of the dataset, as @rgruener  mentioned. e.g.:
```python
rows_rdd = sc.parallelize(ids, 10000)
```
(try playing with 10000)

**Comment by eliorc** [comment]
@rgruener 
The fields which are not in the shape of a number, are arrays of maximum length of 500. I have about 80 fields, which about half of them are arrays, the rest are number and one of them is a string.
About the rows, we are talking about millions.

My Spark version is 2.4.0

@selitvin 
I have 81K~ ids, each one yield on average 7 dictionaries from the `row_generator`. I have tried 81K, 100,000, and 1,000,000. All fail with the same error.

Previously I solved this using TFRecords but since they are so hard to manipulate and slow to access specific rows, we decided to give Petastorm a try.

**Comment by eliorc** [comment]
Solved it,
It seems that it does not manage to to serialize objects bigger than 4GB when transferring the work to the different processes on the `parallelize` call.

Adding the following code before starting a Spark session solved the issue

```
from pyspark import broadcast
import pickle


def broadcast_dump(self, value, f):
    pickle.dump(value, f, 4)  # was 2, 4 is first protocol supporting >4GB
    f.close()
    
    return f.name


broadcast.Broadcast.dump = broadcast_dump
```

**Comment by selitvin** [comment]
Are you using all 80 columns or you are selecting just some of them when you create `Reader`?
At the moment, Petastorm does not handle datasets with relatively small field sizes (we are using it with multi-megabytes images). We are currently working to make it more efficient in the small-field/row scenario.

As a tuning option that may make reading the data somewhat faster, is to create the reader with the following parameters:
```python
make_reader(...., pyarrow_serialize=True, reader_pool_type='process')
```

**Comment by eliorc** [comment]
I am using all of them when I create a `Reader`. Unfortunately I seem to get an error when I try to run `from petastorm import make_reader`, says there is no method called that way.

I have successfully written the Petastorm dataset, but I can't even seem to be able to read it using the following code

```
from petastorm.reader import Reader
from petastorm.tf_utils import make_petastorm_dataset
import tensorflow as tf

reader = Reader('file:///data/hello_world_dataset')  # this is actually my custom dataset
dataset = make_petastorm_dataset(reader)
iterator = dataset.make_initializable_iterator()
tensor = iterator.get_next()

sess = tf.Session()
sess.run(iterator.initializer)
sess.run(tensor)
```

Seems to get stalled infinitely

**Comment by selitvin** [comment]
I suspect that it is simply super slow because of the kind of record structure that you have. As I mentioned it is very different from the data we use Petastorm internally and we'll have a good solution for this in the next release. 

Can you try generating a small dataset: say with just 5 records? Then we can check if the problem is the speed of reading or that it hangs for some other reason.

Some more ideas to try: read directly from Python, so you don't go via Tensorflow at all:
```
reader = Reader('file:///data/hello_world_dataset')
sample = next(reader)
```

Another option to try to speed up reads as for now is:
```
reader = Reader('file:///data/hello_world_dataset', reader_pool=ProcessPool(5, pyarrow_serialize=True))
sample = next(reader)
```

**Comment by eliorc** [comment]
I'm back at TF Records ATM, so I don't have the availability to try the 5 records test.

I did though try the `next(reader)` calls, both of them stall too long too

**Comment by selitvin** [comment]
Got it. I'll update the ticket when we are out with the new version that should handle your case well.

**Comment by shaywinter** [comment]
@selitlevin
Following on above thread, saw you released 0.5.0 10 days ago - is this the version that is supposed to resolve this issue?

**Comment by selitvin** [comment]
Yes, you can now use `make_batch_reader` to open a dataset that was not created by Petastorm. As long as it contains scalar only columns. #273 and #272 extend support for date/timestamp and list types.

**Comment by shaywinter** [comment]
Reading works for me now albeit I noticed that most time is spent on first next(reader), and that this time is linear in data length - could it be that you're scanning all data on first read? Can it be avoided?
In parallel I'm trying make_batch_reader which doesn't have this issue.

**Comment by selitvin** [comment]
Since 0.5.1 was released we support reading [non Petastorm datasets](https://petastorm.readthedocs.io/en/latest/readme_include.html#non-petastorm-parquet-stores). Your column-types, as you have described them, are all native-Parquet compliant (e.g. you don't have multidimensional tensors, or images). I think it would be best to create a Parquet dataset without `materialize_dataset` context manager using pure Parquet types only and then read it using `make_batch_reader`. This should have the most optimal performance.

We don't have a good example that shows the full flow I just described, except for the unit-test [setup code](https://github.com/uber/petastorm/blob/1b082ed9454c296610b5d916c6025b92bb55358a/petastorm/tests/test_common.py#L154), but we plan to add one: #286.

**Comment by selitvin** [comment]
> My datatypes include sequences therefore I cannot use `make_batch_reader`.
`make_batch_reader` does support Parquet array types. If you are using these, your should be fine. If the code still fails for some reason, would be happy to see a call stack so we can address the issue. I would strongly recommend using `make_batch_reader` if it is possible in your case as it provides best performance.

There should not be any difference between process and thread pools in the way they are used. 
Using a process pool is more memory wasteful. Could it be that using multiprocessing triggers violation of some memory-cap? Can it be increased? 

You can also try and minimize memory consumption by reducing the number of processes or the size of the shuffling queue (if you have one set up in the call to `tf_tensors`).

**Comment by YunseokJANG** [comment]
I have a question related to this. 

When I tried to convert very large data (10TB hdf5 file) with the example code in the local server, I think the RDD causes an "out of memory" error. 

random guess) I guess the RDD is processed step by step. To overcome such an issue, I think the sample code should feed the old data to the next RDD, once the number of rows exceeds some specific limit for instance. However, as a newbie to Spark, I failed to find a way to set this logic. Could you let me know how I can change the behavior of it?

**Comment by selitvin** [comment]
Getting an OOM is a very common thing with pyspark - when working with large data, especially images/sensor data. While there could be many different reasons for this, let me try pointing into some directions. I assume, you are running on a single machine with a 'local' spark, i.e.:
```python
spark = SparkSession.builder.config('spark.driver.memory', '2g').master('local[2]').getOrCreate()
```
- Consider increasing `spark.driver.memory` to something much larger (say 64gb or more if you have enough RAM).
- Reduce amount of executors running at the same time by setting: `master('local[1]')`
- Reduce number of rows in a row-group: that's where most of the memory pressure comes from (e,g. 64MB per row-group, as here): `materialize_dataset(spark, ...,... , 64)`
Can you describe the structure of your record? How many fields, what are their sizes?
- Newer versions of parquet-hadoop library has improved memory footprint. Try the newest pyspark.

If you could some code example that could also help...

**Comment by YunseokJANG** [comment]
Thanks for your reply. Actually, I've tried all other bullet points except the "64MB per row-group". However, it seems like pushing every data to RAM for each RDD (if my understanding is correct), and that causes the OOM. In my case, there are 200K items, and one item takes around 20M. I will try the third option if I have some more time. Thanks again!

**Comment by miguelalonsojr** [comment]
@YunseokJANG Did you ever resolve your issue? I'm facing similar problems, OOM, for large datasets, ~100k.

**Comment by YunseokJANG** [comment]
(just in case) @miguelalonsojr I actually no longer handling this issue (changed the dataset to a smaller one). Hope the maintainers resolved this issue.
